gadget-container/gadgets/collector-process/c/ebpf.c

```diff
@@ -0,0 +1,106 @@
+#include <linux/bpf.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+
+struct bpf_iter_meta {
+	__bpf_md_ptr(struct seq_file *, seq);
+	__u64 session_id;
+	__u64 seq_num;
+};
+
+struct bpf_iter__task {
+	struct bpf_iter_meta *meta;
+	struct task_struct *task;
+} __attribute__((preserve_access_index));
+
+struct ns_common {
+	unsigned int inum;
+} __attribute__((preserve_access_index));
+
+struct nsproxy {
+	struct mnt_namespace *mnt_ns;
+} __attribute__((preserve_access_index));
+
+struct task_struct {
+	int pid;
+	int tgid;
+	char comm[16];
+	struct nsproxy *nsproxy;
+} __attribute__((preserve_access_index));
+
+struct mnt_namespace {
+	struct ns_common ns;
+}__attribute__((preserve_access_index));
+
+struct container {
+	char ContainerID[256];
+	char KubernetesNamespace[256];
+	char KubernetesPod[256];
+	char KubernetesContainer[256];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, __u64);
+	__type(value, struct container);
+	__uint(max_entries, 1024);
+} containers SEC(".maps");
+
+
+/* From: tools/lib/bpf/bpf_tracing.h */
+/*
+ * BPF_SEQ_PRINTF to wrap bpf_seq_printf to-be-printed values
+ * in a structure.
+ */
+#define BPF_SEQ_PRINTF(seq, fmt, args...)				    \
+	({								    \
+		_Pragma("GCC diagnostic push")				    \
+		_Pragma("GCC diagnostic ignored \"-Wint-conversion\"")	    \
+		static const char ___fmt[] = fmt;			    \
+		unsigned long long ___param[] = { args };		    \
+		_Pragma("GCC diagnostic pop")				    \
+		int ___ret = bpf_seq_printf(seq, ___fmt, sizeof(___fmt),    \
+					    ___param, sizeof(___param));    \
+		___ret;							    \
+	})
+
+
+SEC("iter/task")
+int dump_task(struct bpf_iter__task *ctx)
+{
+	struct seq_file *seq = ctx->meta->seq;
+	__u32 seq_num = ctx->meta->seq_num;
+	struct task_struct *task = ctx->task;
+
+	if (seq_num == 0)
+		BPF_SEQ_PRINTF(seq, "[\n");
+
+	if (task == (void *)0) {
```

Comment:
Could it ever happen that seq_num is 0 and task is NULL? If this is a possible case, then you'd print `[\n\n]\n` is that intended?... If this is not a possible case then maybe the code can be re-written as:

```
if (task == (void *)0) {
	BPF_SEQ_PRINTF(seq, "\n]\n");
	return 0;
}

if (seq_num == 0) {
	BPF_SEQ_PRINTF(seq, "[\n");
} else {
	BPF_SEQ_PRINTF(seq, ",\n");
}
```

Reply:
> Could it ever happen that seq_num is 0 and task is NULL?

I think it's not possible for this iterator on processes: it would mean there are zero processes running on this Linux machine. If there was no processes, nobody would be able to call this iterator.

However, I think it is worth fixing this properly because:
* We might create a collector-tcp gadget (based on [bpf_iter_tcp4.c](https://github.com/torvalds/linux/blob/3dbdb38e286903ec220aaf1fb29a8d94297da246/tools/testing/selftests/bpf/progs/bpf_iter_tcp4.c) or a collector-udp gadget (based on [bpf_iter_udp4.c](https://github.com/torvalds/linux/blob/3dbdb38e286903ec220aaf1fb29a8d94297da246/tools/testing/selftests/bpf/progs/bpf_iter_udp4.c)) and in this case, it's possible to have zero tcp or zero udp sockets. If we write the correct code pattern for collector-process, then it can just work for the other gadgets.
* We could implement filters: return a subset of processes.